docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,753 @@
+    /**
+     * The original download request given to the Transfer Manager.
+     */
+    DownloadObjectRequest downloadRequest();
+
+    /**
+     * The request sent to S3 for this object. This is empty if downloading a presigned URL.
+     */
+    Optional<GetObjectRequest> objectRequest();
+}
+
+/**
+ * Progress listener for a Transfer.
+ * <p>
+ * The SDK guarantees that calls to {@link #transferProgressEvent(EventContext)}
+ * are externally synchronized.
+ */
+public interface TransferProgressListener {
+    /**
+     * Called when a new progress event is available for a Transfer.
+     *
+     * @param ctx The context object for the given transfer event.
+     */
+    void transferProgressEvent(EventContext ctx);
+
+    interface EventContext {
+    }
+
+    interface Initiated extends EventContext {
+        /**
+         * The transfer this listener associated with.
+         */
+        Transfer<?> transfer();
+    }
+
+    interface BytesTransferred extends Initiated {
+        /**
+         * Average number of bytes transferred per second for this transfer.
+         */
+        long averageTransferBytesPerSecond();
+
+        /**
+         * The transfer request for the object whose bytes were transferred.
+         */
+        TransferObjectRequest objectRequest();
+
+        /**
+         * The number of bytes transferred for this event.
+         */
+        long bytes();
+
+        /**
+         * The total length of the object.
+         */
+        long objectLength();
+
+        /**
+         * If the transfer of the given object is complete.
+         */
+        boolean complete();
+    }
+
+    interface Completed extends Initiated {
+    }
+
+    interface Cancelled extends Initiated {
+    }
+
+    interface Failed extends Initiated {
+        /**
+         * The error.
+         */
+        Throwable error();
+    }
+}
+
+/**
+ * Hints to help optimize an object download.
+ */
+public interface DownloadHints {
+    /**
+     * The size of the object to be downloaded. In parallel download situations
+     * this allows the Transfer Manager to avoid making an extra request to S3
+     * to find the size of the object in order to determine proper object
+     * partitioning.
+     */
+    long objectLength();
+}
+
+/**
+ * A download transfer of a single object from S3.
+ */
+public interface DownloadObject extends Transfer<DownloadObject> {
+    @Override
+    CompletableFuture<DownloadObject> completionFuture();
+
+    @Override
+    DownloadObjectState pause();
+}
+
+
+/**
+ * An upload transfer of a single object to S3.
+ */
+public interface UploadObject extends Transfer<UploadObject> {
+    @Override
+    CompletableFuture<UploadObject> completionFuture();
+
+    @Override
+    UploadObjectState pause();
+}
+
+/**
+ * Represents the transfer of one or more objects to or from S3.
+ */
+public interface Transfer<T extends Transfer> {
+
+    CompletableFuture<T> completionFuture();
```

Comment:
Should we do `CompletableFuture<?>` to avoid needing the parameterization of the transfer? It feels like customers would always use `Transfer<?>` anyway.

Reply:
I'm in favor of `CompletableFuture<CompletedDownloadObject>` given above instead. I think it'd be nice to be able to expose information about the transfer after it has completed.

Reply:
Since this is the base class, maybe `CompletableFuture<? extends CompletedTransfer>`?

Reply:
+1